### ISSUE TITLE
Complete OCF

### ISSUE THREAD

**Issue by hs0225** [Samsung/iotjs#566]
Implement OCF(Open Connectivity Foundation)

- [ ] core
  - [ ] initialize
  - [ ] get_properties
  - [ ] setup_properties
  - [ ] free_ocf

- [ ] server
  - [ ] initialize
  - [ ] create_resource
  - [ ] create_response
  - [ ] create_request
  - [ ] get_handler
  - [ ] put_handler
  - [ ] delete_handler
  - [ ] post_handler
  - [ ] notify
  - [ ] register
 
- [ ] client
  - [ ] initialize
  - [ ] create_handler
  - [ ] get_handler
  - [ ] create_url
  - [ ] find_device
  - [ ] find_resource
  - [ ] create_resource
  - [ ] add_resource
  - [ ] create
  - [ ] update
  - [ ] delete

**Comment by LaszloLango** [comment]
What is the motivation of re-implementing instead of using iotivity-constrained?

**Comment by RedCarrottt** [comment]
I have a great interest in OCF support of IoT.js since I have a plan to make a graduate school lecture covering IoT.js and IoTivity.
Does the OCF IoT.js module just use existing IoTivity or IoTivity-constrained as a library?

**Comment by yichoi** [comment]
@LaszloLango Basically, I am going to use iotivity-constrained however As far as I know, iotivity-constrained does not support nuttx and our target (you might aware of this). I am thinking whether we should implement iotivity-constrained for nuttx and our target. Alternatively, I could assume that there will be implementation of iotivity-constrained for nuttx and our target then only support Linux for a while.

**Comment by rzr** [comment]
As some of you already know, parts (csdk) of iotivity 1.2-rel and 1.3-rel are already in TizenRT (so could be easly ported ported to nuttx), iotvivity-constrained works too on Tizen or Linux, but I am not aware about any attempt to port to nuttx/tinyara.

Also check related projects: zephyr.js (with IoTivity-contrained) and iotivity-node (based on regular IoTivity), maybe they can be aligned (at least licenses are same).

**Comment by haesik** [comment]
@rzr Yes, I know iotivity 1.2 and 1.3 are already in the TizenRT and Tizen also support iotivity.
But I think the device-cloud model is more useful in IoT scenario, so we are focused on supporting SmartThings internally(that is also based on iotivity). Consequentially, we don't have enough human resource to support iotivity now.
